src/hierarchies/base_hierarchy.h

```diff
@@ -34,30 +34,33 @@ class BaseHierarchy {
   std::set<int> cluster_data_idx;
   int card = 0;
   double log_card = stan::math::NEGATIVE_INFTY;
-
+  //!
   virtual void update_summary_statistics(const Eigen::VectorXd &datum,
+                                         const Eigen::VectorXd &covariate,
                                          bool add) = 0;
+  //!
+  virtual void save_posterior_hypers() = 0;
 
  public:
-  virtual void add_datum(const int id, const Eigen::VectorXd &datum);
-
-  virtual void remove_datum(const int id, const Eigen::VectorXd &datum);
-
+  //! Adds a datum and its index to the hierarchy
+  void add_datum(const int id, const bool save_params,
```

Comment:
I would rename `save_params` as `update_params`, also we could have the signature
```
add_datum(const int id, const Eigen::VectorXd& datum const bool save_params, bool update_params = false, const Eigen::VectorXd &covariate = Eigen::VectorXd(0))
```

Reply:
Sure. Is there a reason why you switched the bool and the datum? I was trying to keep all data information close to each other.